Chatkit/Redux/Reducers/MasterReducer.swift

```diff
@@ -0,0 +1,56 @@
+
+extension Reducer {
+    
+    struct Master: Reducing {
+        
+        // MARK: - Types
+        
+        typealias ActionType = Action
+        typealias StateType = MasterState
+        typealias DependenciesType =
+            HasUserReducer
+            & HasRoomListReducer
+            & HasUserSubscriptionInitialStateReducer
+            & HasUserSubscriptionAddedToRoomReducer
+            & HasUserSubscriptionRemovedFromRoomReducer
+            & HasUserSubscriptionRoomUpdatedReducer
+            & HasUserSubscriptionRoomDeletedReducer
+            & HasUserSubscriptionReadStateUpdatedReducer
+        
+        // MARK: - Reducer
+        
+        static func reduce(action: ActionType, state: StateType, dependencies: DependenciesType) -> StateType {
+            
+            if let action = action as? InitialStateAction {
+                return dependencies.initialStateUserSubscriptionReducer(action, state, dependencies)
+            }
+            else if let action = action as? AddedToRoomAction {
+                return dependencies.userSubscriptionAddedToRoomReducer(action, state, dependencies)
+            }
+            else if let action = action as? RemovedFromRoomAction {
+                return dependencies.userSubscriptionRemovedFromRoomReducer(action, state, dependencies)
+            }
+            else if let action = action as? RoomUpdatedAction {
+                return dependencies.userSubscriptionRoomUpdatedReducer(action, state, dependencies)
+            }
+            else if let action = action as? RoomDeletedAction {
+                return dependencies.userSubscriptionRoomDeletedReducer(action, state, dependencies)
+            }
+            else if let action = action as? ReadStateUpdatedAction {
+                return dependencies.userSubscriptionReadStateUpdatedReducer(action, state, dependencies)
+            }
+            
+            return state
+        }
+        
+    }
+    
+}
+
+// MARK: - Dependencies
+
+protocol HasMasterReducer {
```

Comment:
I've noticed that you've moved all the `protocol Has*` to the bottom of the file where-as I would expect to see relevant protocols for a given class at the top.  (for example you'd normally define `Delegate` protocols above the class itself).

This also keeps things consistent with the pre-existing `InstanceLocator`, `Store` and `StoreBroadcaster`.

Reply:
This is a matter of personal preference. One could argue that when a file is named `Name.swift` than the most important element of this file (the class/struct/enum/etc. `Name`) should be placed on top of the file. Such approach also allows pragma marks to be places in a way which then nicely integrates with Xcode's navigation.

Reply:
Sure, everything can be attributed to personal preference.  I'm happy to be flexible on these minor details but I do like things to be consistent.  So I think we either need to update all the existing `protocol Has*` moving them to the bottom or move these back to the top. 

Not terribly fussed either way but last time we didn't share opinion on a given personal preference the decision was made not to change things for change sake and to continue with the approach that had set a precedent. 

I'll leave it up to you.  I'm happy to approve either way.